proposed/before_bundle_build_hook.md

```diff
@@ -0,0 +1,49 @@
+
+In addition to signing a similar difficulties may affect other scenarios that rely on post-build modification of build artefacts. For example obfuscation and instrumentation.
+
+## Current state (as of .NET 6)
+
+The problem of signing applies primarily to files that are extracted to the file system as a part of execution. For the assemblies that are not extracted the problem is generally not a concern. 
+With transitioning to superhost technology on all platforms in 6.0 a need to extract files is greatly reduced as native binaries that belong to the dotnet platform are no longer extracted and in fact cannot be extracted, since they are monolithically linked with the app. 
+
+The files that still can be extracted:
+-	Native dependencies included via a nuget references. (this is, technically, the only truly supported scenario to include a native dependency)
+It could be expected that such files are signed prior to the build.
+
+-	Dependencies that are forced to be extracted by the user.  
+There are several possible reasons for the user to force extraction of all files and there is a way to force extraction of all dependencies, including managed assemblies that otherwise could run in-place.
+Note that these files include the app dll itself as well as `.r2r.dll` in a case of composite publishing. Both would not present themselves to the user in the process of building.
+This is the case where extraction is a very observable feature and it is plausible the extraction as a feature will be carried forward in the future versions.
+I believe this is the case that causes the most inconveniences.
+
+The abscense of documented solutions in this area forces users to implement solutions that rely on internal implementation details.  
+For example: [`CollectBundleFilesToSign` target](https://github.com/dotnet/diagnostics/blob/18a8e986ddf28ee3c950b5acf00abc652d87be74/src/Tools/Directory.Build.targets).  
+
+## Proposal for .NET 6
+
+Provide a build-time hook to implement signing logic. The use pattern is expected to be similar to `CollectBundleFilesToSign`, but without relying on undocumented behavior.  
+We will provide:
+-	A target `BeforeBundle` that will be called between `_ComputeFilesToBundle` and `GenerateBundle`
+-	An ItemGroup `FilesToBundle` containing all files that will be passed to `GenerateSingleFileBundle`
+-	A Property `AppHostFile` that will specify the apphost.  
+The reason for this is that the host will be in the list of files to bundle, but generally need not be signed.
+
+It is possible that the signing tool will need to copy files in the process of signing. That could happen if the file is a shared item not owned by the build, for example the file comes from a nuget cache. In such case it is expected that the tool will modify the path of the corresponding `FilesToBundle` item to point to the signed copy.
+
+## Alternative approaches considered:
+
+-	Unbundle/Rebundle support
+
+Provide a way to unbundle/rebundle a singlefile app.  
```

Comment:
There's one scenario where this is valuable

- Apps that need to get signed on different platforms: dylib requiring signatures on macOS, dll's needing signing but that can't be done on the machine (like we do for security and compliance reasons).
- The general tool-needed machine marshaling issue - a lot of things need an apple device such as `codesign` and notarization/hardening, or a windows device because the tool operates on Windows only, etc...

Reply:
+1 that this should be mentioned in the doc. We decided to implement a simple solution for .NET 6 and collect more feedback before investing into a more complex solution which could support "async" processing of the files in the bundle.

Reply:
Right. The proposal assumes that signing can be logically done as a part of a build, even if files need to be sent to a service or copied around.  
If build and sign machines are different, the copying would need to happen regardless of the solution, but not having to do it while the build is waiting may have some advantages. I can imagine situations when the service is very slow, unreliable or need to happen on a disconnected machine and holding up the build could become a problem. How big such a problem is in practice - it would be useful to listen to feedback here.

I will mention this concern.